Interpret overhaul, adding MarkupSchema

w0nche0l opened the pull request:
Converted currently existing code to use MarkupBlockSchema instead of a raw dictionary. 
Created MarkupBlockSchema. 
Created a new block_schemas.py to avoid circular dependency errors; added BlockModes enum class to keep track of mode strings in centralized location. 
See https://github.com/codalab/codalab-worksheets/pull/427 for front-end changes.


WIP: Other blocks. 

Final goal of this PR: to use BlockSchemas for all interpreted items.

w0nche0l committed:
[WIP] do not commit

w0nche0l committed:
[WIP] do not commit

w0nche0l committed:
[WIP] converted markup to use BlockSchemas

w0nche0l committed:
[WIP] converted markup to use BlockSchemas

w0nche0l committed:
[WIP] converted markup to use BlockSchemas

w0nche0l committed:
[WIP] converted markup to use BlockSchemas

w0nche0l committed:
fixing accidental deletion

w0nche0l committed:
fixing accidental deletion

percyliang commented on codalab/lib/worksheet_util.py:
```diff
@@ -617,11 +618,11 @@ def raise_genpath_usage_error():
+                    interpreted_items.append(
+                        MarkupBlockSchema().load({
+                            'id': len(interpreted_items),
+                            "text": "ERROR: cannot access bundle",
```
be consistent: use single quotes

w0nche0l replied:
Oops @ being consistent; I thought we supposed to be using double quotes throughout the codebase?

percyliang commented on codalab/lib/worksheet_util.py:
```diff
@@ -617,11 +618,11 @@ def raise_genpath_usage_error():
-                        'interpreted': 'ERROR: cannot access bundle',
-                        'properties': {},
-                    })
+                    interpreted_items.append(
```
Should call these `blocks` instead of `interpreted_items` to avoid confusion...

w0nche0l replied:
sounds good!

percyliang replied:
Still need to do this, right?

w0nche0l replied:
yeah, wanted to simplify the diff view by doing this last

percyliang commented on codalab/rest/block_schemas.py:
```diff
@@ -0,0 +1,96 @@
+"""
+Marshmallow schemas for serializing resource dicts into JSON API documents, and vice-versa.
```
Should say something about which schemas (worksheet blocks).

Naming: at the top-level, should be specific and use `worksheet_block` (which goes for the file name).

percyliang commented on codalab/rest/block_schemas.py:
```diff
@@ -0,0 +1,96 @@
+
+# Enum that represents different modes for a block.
+class BlockModes:
+    MARKUP_MODE = "markup_block"
```
Can we just make the Python variable match the actual string?

percyliang commented on codalab/rest/block_schemas.py:
```diff
@@ -0,0 +1,96 @@
+STATUS_STRINGS = ("unknown", "ready", "not_found", "no_permission")
+
+
+class FetchStatusSchema(fields.Field):
```
Document

percyliang commented on codalab/rest/block_schemas.py:
```diff
@@ -0,0 +1,96 @@
+              SEARCH_MODE)
+
+
+STATUS_STRINGS = ("unknown", "ready", "not_found", "no_permission")
```
Need to think more carefully about whether this is the right set of status strings...

For example, if I don't have permission, should I get a 'not_found' or 'no_permission' (which reveals the existence of certain resources even if I don't have access)...

w0nche0l replied:
I think it is okay to have a distinction between not_found and no_permission, since we already do that for resources in our current front-end.

w0nche0l replied:
What is unknown? Just like a generic catch-all?

w0nche0l replied:
Should there be a "fetching" status that represents something that is currently trying to be fetched?

percyliang replied:
I think `unknown` is supposed to be that.  I don't think it should be called `fetching` because it should apply when no fetching request has been made.  Maybe `pending`?

percyliang commented on codalab/rest/interpret.py:
```diff
@@ -316,14 +317,15 @@ def error_data(mode, message):
         mode = item['mode']
-        data = item['interpreted']
-        properties = item['properties']
+        # TODO: remove these fields entirely, using fields unique to each mode instead.
```
What's the migration plan?  Why keep these here if we plan to use the block-schema-specific fields...

w0nche0l replied:
I'm going to update this PR to remove these entirely, they're in there currently to keep every other block type working while I test.

w0nche0l committed:
modified to support html, contents, and image

w0nche0l committed:
modified to support html, contents, and image

w0nche0l commented on codalab/rest/interpret.py:
```diff
@@ -296,7 +296,6 @@ def head_target(target, max_num_lines, replace_non_unicode=False):
 
 
 # Default number of lines to pull for each display mode.
-DEFAULT_CONTENTS_MAX_LINES = 10
```
Moved this constant to worksheet_util.py

w0nche0l commented on codalab/rest/interpret.py:
```diff
@@ -323,40 +322,55 @@ def error_data(mode, message):
                 data = (header, contents)
-            elif mode == 'contents':
+
+                # TODO: remove this and use new schemas
```
All of these are marking things that I haven't migrated.

w0nche0l commented on codalab/rest/worksheet_block_schemas.py:
```diff
@@ -0,0 +1,105 @@
+
+class BundleHTMLBlockSchema(BundleBlockSchema):
+    mode = fields.Constant(BlockModes.html_block)
+    max_lines = fields.Constant(sys.maxint)
```
Currently, the html display function is broken on production. This is a hack to make sure the entire HTML file gets displayed, but I'm not a huge fan. However, file_util:summarize_file explicitly asserts to make sure that there is some limit on the number of lines. Help!

percyliang replied:
What about deleting the html display functionality?  Does it actually get used?  For rich enough HTML, one would want to display it in another window anyway because we can't render arbitrary HTML with Javascript anyway in an embedded way...  Haven't thought this through but just bringing it up.

w0nche0l commented on codalab/rest/worksheet_block_schemas.py:
```diff
@@ -0,0 +1,105 @@
+
+# Enum that represents different modes for a block.
+class BlockModes:
+    markup_block = 'markup_block'
```
Renamed to match strings.

w0nche0l commented on codalab/rest/worksheet_block_schemas.py:
```diff
@@ -0,0 +1,105 @@
+              search_block)
+
+
+STATUS_STRINGS = ('unknown', 'pending', 'ready', 'not_found', 'no_permission')
```
Added pending status (not used yet), to be used when the front-end takes a block at unknown status and sends it off to be refined.

w0nche0l commented on codalab/lib/worksheet_util.py:
```diff
@@ -615,28 +624,52 @@ def raise_genpath_usage_error():
+                    interpreted_items.append(BundleHTMLBlockSchema().load({
+                        'target_genpath': target_genpath,
+                        'bundle_info': copy.deepcopy(bundle_info),
+                        'status': {
```
Maybe I should define a constant for initialization of statuses?

percyliang replied:
Yes, I think it's worth defining a bunch of helpers to reduce code duplication as long as it doesn't introduce layers of indirection that make it hard to understand.

percyliang commented on codalab/rest/worksheet_block_schemas.py:
```diff
@@ -0,0 +1,105 @@
+    image_block = 'image_block'
+    graph_block = 'graph_block'
+    wsearch_block = 'wsearch_block'
+    search_block = 'search_block'
```
I think we decided that `search` was not going to be a separate type of block, but a mechanism for retrieving a list of bundles; specifying a list of bundles verbatim is another mechanism.

w0nche0l committed:
finishing up refactors to use schemas; missing search/wsearch

w0nche0l committed:
finishing up refactors to use schemas; missing search/wsearch

w0nche0l committed:
WIP new BundlesSpecSchema

w0nche0l committed:
WIP new BundlesSpecSchema

w0nche0l committed:
adding bundlesspecschema

w0nche0l committed:
adding bundlesspecschema

percyliang commented on codalab/rest/worksheet_block_schemas.py:
```diff
@@ -0,0 +1,175 @@
+    error_message = fields.String()
+
+    @staticmethod
+    def get_initial_status():
```
rename to `get_unknown_status`

w0nche0l committed:
[WIP] adding new schemas for search

w0nche0l committed:
[WIP] adding new schemas for search

w0nche0l committed:
finalized search implementation

w0nche0l committed:
finalized search implementation

w0nche0l committed:
refactoring how search/wsearch works, and also unifying worksheet logic

w0nche0l committed:
refactoring how search/wsearch works, and also unifying worksheet logic

w0nche0l committed:
fixing syntax errors

w0nche0l committed:
fixing syntax errors

percyliang commented on codalab/lib/bundle_cli.py:
```diff
@@ -2339,37 +2339,31 @@ def do_print_command(self, args):
-                if mode == 'contents':
-                    maxlines = properties.get('maxlines')
+            if mode == 'markup_block' or mode == 'contents_block':
+                bundle_info = item['bundles_spec']['bundle_infos'][0]
```
If markup, this shouldn't exist, right?

percyliang replied:
Probably these two modes should just be separate branches of the if.

w0nche0l replied:
yeah, will definitely change this.

percyliang commented on codalab/lib/worksheet_util.py:
```diff
@@ -36,25 +36,43 @@
 CURRENT_WORKSHEET = '.'
 
-# Types of worksheet items
+# Types of raw worksheet items
```
say '(raw)', since by default worksheet items are raw now...

percyliang commented on codalab/lib/worksheet_util.py:
```diff
@@ -679,45 +712,59 @@ def raise_genpath_usage_error():
+
+            interpreted_items.append(GraphBlockSchema().load({
+                'trajectories': trajectories,
+                'bundles_spec': BundleUUIDSpecSchema().load(BundleUUIDSpecSchema.create_json([bundle_infos[0][1]])).data,
```
Keep the comment that we're taking only the first bundle

percyliang commented on codalab/rest/interpret.py:
```diff
@@ -183,6 +186,12 @@ def fetch_interpreted_worksheet(uuid):
     # Fetch items.
     worksheet_info['raw'] = get_worksheet_lines(worksheet_info)
 
+    # Replace searches with raw items.
```
Do you think you'll be able to unify this with `fetch_worksheet` from `worksheets.py`?

w0nche0l replied:
I'm not sure I understand this question, fetch_worksheet only cares about the raw items in the worksheet, whereas this is replacing any raw items that contain search directives with the result of the search? For example, if my worksheet just contained 
"% search .limit=2"
I would not want the worksheet to seem like it actually contained
[bundle1]{0xw50u5oijwed} 
[bundle2]{0xoih0w3523n5}

percyliang replied:
I just meant that there would be one REST endpoint that would return both the (raw) items as well as the blocks.  Otherwise there's some code duplication in two places, and there's inconsistent handling of owner, for example.

w0nche0l replied:
Ah okay I see what you mean, I'll take a look.

w0nche0l replied:
There's not a very clean way to unify this, I think this should be separated into its own diff after this is merged.

percyliang commented on codalab/rest/interpret.py:
```diff
@@ -584,3 +544,39 @@ def resolve_items_into_infos(items):
+
+def expand_raw_item(raw_item):
+    """
+    Some raw items must be expanded into more raw items.
```
Be more specific about how it's just search items...

percyliang commented on codalab/rest/interpret.py:
```diff
@@ -183,6 +186,12 @@ def fetch_interpreted_worksheet(uuid):
+    # This needs to be done before get_worksheet_lines because this replaces
+    # user-written raw items.
+    worksheet_info['items'] = list(chain.from_iterable(
+        [expand_raw_item(raw_item) for raw_item in worksheet_info['items']]))
```
This really worries me...`items` is now a list of a list of items?  It'd be nice if `items` really just corresponded to the raw items.  I thought the search directives would only influence the blocks.

w0nche0l replied:
No, the chain.from_iterable flattens the list of list of items from the list comprehension into a list of items. Before I changed anything worksheet_info['items'] was populated with raw items, and then was replaced with the interpreted items. I can make it so that this is completely separated into worksheet_info['raw_items'], worksheet_info['replaced_raw_items'], and worksheet_info['interpreted_blocks'], if that makes more sense. 

I think this actually fits really nicely into how our parser/interpreter works currently, because a search is an alias for a list of bundles (or worksheets, in the case of wsearch), and this means that we don't have to include extra logic for every single block on how to handle searches.

percyliang replied:
Well, one question is whether we want the conversion from items to blocks to be oblivious of search directives.  Certainly doing this replacement simplifies the blocks logic, but it does introduce 3 objects to reason about (raw items, replaced items, and blocks) rather than 2.  Also, it precludes us from later trying to lazily resolve the search directives during interpretation time rather than all at once.

w0nche0l replied:
I don't think it gives us 3 objects to reason about, since replaced items are identical to raw items from the perspective of the raw -> block interpreter. 

Also, as we discussed before, search directives are important to the codalab architecture, but also used very sparsely, mostly in the homepage and to perform debugging. Therefore, I think it would be much more efficient to add logic for lazily resolving other directives, such as image and table, since that functionality would apply to all worksheets. Also, those gains would be realized for searches, since searches resolve into table blocks anyway.

percyliang replied:
As long as only raw items and blocks are exposed to the frontend, then I'm happy.  If the replaced items can be encapsulated in the raw items => blocks conversion (even though we do break it up further, which is great)...

I don't understand your point about searches - can you please rephrase?

Are we committing to resolving all the search directives whenever we fetch a worksheet?  This seems to make certain things off limits - like being able to lazily fetch a very long list of bundles or worksheets dynamically...  Maybe it's fine to do that, but it seems worth thinking through the consequences.

w0nche0l replied:
Yes, currently the frontend is completely unaware of the raw items. It renders purely based on the blocks. 

Sorry for being unclear; I think what I was trying to say is that the plan is already to lazily resolve the contents of display directives that require more than just the bundle metadata, such as images, contents, and genpaths in tables. These all require disk reads, and are what contribute most to the slow load times of large and complex worksheets. For example, if a table of 10 items wants the first line of the file results.txt, that will hit the file system 10 separate times. I was under the impression that this refactor is laying the grounds for lazily loading these disk-intensive tasks. 

Meanwhile, search is definitely not a common tool for most worksheets, it is mostly used as a diagnostic tool for us to build dashboards such as the current homepage. Additionally, it will gain the same performance gains that I described above. 

We are committing to resolving all of the search directives whenever we fetch a worksheet, but that was already being done previously, so it wouldn't make anything slower than it already is now. I'm having a hard time thinking of a use case for lazily fetching a very long (>1000) list of bundles dynamically.

percyliang commented on codalab/rest/interpret.py:
```diff
@@ -306,134 +314,86 @@ def resolve_interpreted_items(interpreted_items):
+                    if mode == BlockModes.contents_block:
+                        item['lines'] = head_target((item['bundles_spec']['bundle_infos'][0]['uuid'], item['target_genpath']), item['max_lines'], replace_non_unicode=True)
+                    elif mode == BlockModes.html_block:
+                        item['html_lines'] = head_target((item['bundles_spec']['bundle_infos'][0]['uuid'], item['target_genpath']), item['max_lines'])
```
html_lines doesn't make much sense since truncated HTML isn't valid.  Probably better to just have `html` and slurp up the entire file (like how we treat images).

I thought we were going to deprecate HTML anyway?

w0nche0l replied:
Yeah we were going to deprecate HTML, should I do it in this change? was just replicating functionality for this change and then take it out in a separate one.

percyliang commented on codalab/rest/interpret.py:
```diff
@@ -306,134 +314,86 @@ def resolve_interpreted_items(interpreted_items):
-                except ValueError:
-                    raise UsageError("maxlines must be integer")
-
+                    item['status']['code'] = 'not_found'
```
probably should have constants defined for the various codes...

w0nche0l replied:
yeah, sounds good.

percyliang commented on codalab/rest/interpret.py:
```diff
@@ -306,134 +314,86 @@ def resolve_interpreted_items(interpreted_items):
-            return (('ERROR',), [{'ERROR': message}])
-        else:
-            return [message]
+    def error_data(item_index, message):
```
Why make this mutable?  If you really need it to mutate, then rename to `set_error_data`.

w0nche0l replied:
I got the impression that this entire function is mutating the contents of interpreted_items anyway, I can make a non-mutating version that creates a copy first and fills it in.

percyliang reviewed (changes requested):
Main comment is regarding how expanding raw items is handled.

w0nche0l committed:
addressing comments and fixing bugs

w0nche0l committed:
addressing comments and fixing bugs

w0nche0l committed:
fixing errors in tests

w0nche0l committed:
fixing errors in tests

w0nche0l committed:
fixing graph

w0nche0l committed:
fixing graph

w0nche0l committed:
Removing HTML

w0nche0l committed:
Removing HTML

w0nche0l committed:
adding support for print by adding method for rest client

w0nche0l committed:
adding support for print by adding method for rest client

bkgoksel commented on codalab/lib/worksheet_util.py:
```diff
@@ -679,45 +709,61 @@ def raise_genpath_usage_error():
+            blocks.append(GraphBlockSchema().load({
+                'trajectories': trajectories,
+                'bundles_spec': BundleUUIDSpecSchema().load(BundleUUIDSpecSchema.create_json([bundle_infos[0][1]])).data, # Only show the first one for now
+                # 'bundles_spec': BundleUUIDSpecSchema().load(BundleUUIDSpecSchema.create_json(
```
Do we know what this is or is it just legacy commented out code?

bkgoksel reviewed (approved):
I'm not super familiar with the details of how everything works but the code looks good to me.

bkgoksel committed:
Correctly catch not found errors in interpret

bkgoksel committed:
Correctly catch not found errors in interpret

bkgoksel committed:
Completely remove remaining HTML references in interpret

bkgoksel committed:
Completely remove remaining HTML references in interpret

Commit:
The Great Worker Reconciliation (#909)

New worker architecture, centralized run state tracking, resumable runs and more...

bkgoksel committed:
Merge branch 'dennis-interpret-overhaul' of github.com:codalab/codalab-cli into dennis-interpret-overhaul
